docs/rfcs/017-pageserver-op-atomicity.md

```diff
@@ -0,0 +1,153 @@
+# Durability and atomicity of tenant/timeline operations
+
+The pageserver has 8 tenant/timeline operations, listed below.  In
+addition to that, data can be appended to a timeline by WAL receiver,
+pages can be requested by the compute node, and tenant/timeline status
+can be queries through the mgmt API. But these are the operations that
+modify state in pageserver or in S3, and need to worry about crash
+safety.
+
+To make these operations atomic and recoverable, let's introduce a new
+"tenant index file", called `tenant.json`. For each tenant, there is
+one tenant index file, and it contains a list of all timelines for
+that tenant:
+
+{
+  tenant_id: a93a94724945e95e1a0c448004ece2ec
+
+  timelines: [
+    { timeline_id: "9979cd302340a058606473912651f27f",
+	  ancestor_id: ""
+	  ancestor_lsn: "0/0"
+    },
+    { timeline_id: "f0a6f3372d273dd9ca3480d19e6b565c",
+	  ancestor_id: "9979cd302340a058606473912651f27f"
+	  ancestor_lsn: "1/1698C48"
+	},
+  ]
+}
+
+The file only contains the immutable metadata of each timeline, like
+the point it was branched from. The changing parts, like
+disk_consistent_lsn, are still stored in the per-timeline metadata
+file.
+
+This file allows us to resolve some ambiguous situations, like
+remembering that a tenant exists when it doesn't have any timelines.
+It also allows us to quickly fetch the list of all timelines of a
+tenant, without having to perform S3 LIST operations.
+
+Below is a brief description of all the pageserver tenant/timeline
+operations, and how the steps of creating/deleting local files or
+directories and uploads to S3 are performed. The steps are listed in
+such an order that each operation can be sanely recovered or aborted,
+if the pageserver crashes while the operation is being perfromed.
+
+## Create tenant
+
+Create an empty tenant. It doesn't have any timelines initially.
+
+1. Create local tenant-directory with .temp extension
+2. Create tenant.json file in the directory, with a special flag
+   indicating that the tenant-creation is in progress
+3. Rename the local tenant directory in place
+4. Upload the tenant.json file to S3, without the flag
+5. Update the local file, removing the flag
+
+At pageserver startup, if we see a tenant.json file with the special
+flag, check if the tenant exists in S3. If not, remove the local directory.
+Otherwise remove the flag from local file.
+
+## Create timeline
+
+Create a timeline for a tenant, as result of running initdb.
+
+1. create timeline directory locally, with .temp extension
+2. run initdb, creating the initial set of layers
+3. upload all layer files to S3
```

Comment:
Has this be done synchronously?
Right now, we _schedule_ layer uploads, but do not wait for them on timeline creation. This eventually leads us to updating metadata on the remote storage, but does not happen instantly.

Since presumably we're _moving_ away most of the timeline-related constant fields out of `metadata` to the new tenant index, there's not much to upload at step `4.` either for the newly created timeline.

Same questions are valid for branching process.

Reply:
It can be done asynchronously, meaning that the management API call can return HTTP 200 to the caller as soon as the uploads are scheduled. As long as we do the steps in right order.

Reply:
let me clarify, the problem can arise because we do not remove temp extension from directory before returning 200 to the client, and because of the late rename combined with temp directories removal on startup we can loose created timeline